### ISSUE TITLE
EDDTableFromHttpGet graph error

### ISSUE THREAD

**Issue by emmerbodc** [BobSimons/erddap#32]
Hi Bob, 

We are trying a new method for ingesting NRT data - https://coastwatch.pfeg.noaa.gov/erddap/download/setupDatasetsXml.html#EDDTableFromHttpGet

We have loaded some data in ERDDAP but noticed that when we used the graph feature an error occurs and the dataset is subsequently dropped from ERDDAP. 

We are not sure if we have a XML issue in our setup or if this is a bug? 

Details below. 
Thanks,
Emma
 

The log errors
``` shell
gov.noaa.pfel.erddap.dataset.WaitThenTryAgainException: There was a (temporary?) problem. Wait a minute, then try again. (In a browser, click the Reload button.)
(Cause: com.cohort.util.SimpleException: ERROR while reading /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl: columnName=timestamp not found in longitude,latitude,airTemp,time.)
 at gov.noaa.pfel.erddap.dataset.EDDTableFromFilesCallable.call(EDDTableFromFilesCallable.java:154)
 at gov.noaa.pfel.erddap.dataset.EDDTableFromFilesCallable.call(EDDTableFromFilesCallable.java:31)
 at gov.noaa.pfel.erddap.dataset.EDDTableFromFiles.getDataForDapQuery(EDDTableFromFiles.java:3696)
 at gov.noaa.pfel.erddap.dataset.EDDTable.getTwawmForDapQuery(EDDTable.java:1663)
 at gov.noaa.pfel.erddap.dataset.EDDTable.saveAsImage(EDDTable.java:3861)
 at gov.noaa.pfel.erddap.dataset.EDDTable.respondToDapQuery(EDDTable.java:3117)
 at gov.noaa.pfel.erddap.Erddap.doDap(Erddap.java:4362)
 at gov.noaa.pfel.erddap.Erddap.doGet(Erddap.java:556)
Caused by: com.cohort.util.SimpleException: ERROR while reading /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl: columnName=timestamp not found in longitude,latitude,airTemp,time.
 at gov.noaa.pfel.erddap.dataset.EDDTableFromHttpGet.readFile(EDDTableFromHttpGet.java:370)
 at gov.noaa.pfel.erddap.dataset.EDDTableFromHttpGet.lowGetSourceDataFromFile(EDDTableFromHttpGet.java:322)
 at gov.noaa.pfel.erddap.dataset.EDDTableFromFiles.getSourceDataFromFile(EDDTableFromFiles.java:3029)
 at gov.noaa.pfel.erddap.dataset.EDDTableFromFilesCallable.call(EDDTableFromFilesCallable.java:109)
 ... 32 more
```
``` shell
#0 get data from /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl
  Table.readJsonlCSV(/local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl) done. nColumns=4 nRows=13 TIME=1ms
>> EDDTableFromHttpGet inbound_NRT_5f88_a3eb_59e0 nThreads=1 thread=https-openssl-apr-8443-exec-6 task=0: caught while reading file=/local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl: com.cohort.util.SimpleException: ERROR while reading /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl: columnName=timestamp not found in longitude,latitude,airTemp,time.
  Table.readJsonlCSV(/local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl) done. nColumns=4 nRows=13 TIME=1ms
  Table.readFlatNc /local/tomcat/content/erddapBigDir/dataset/e0/inbound_NRT_5f88_a3eb_59e0/badFiles.nc finished. nColumns=3 nRows=1 TIME=1ms
inbound_NRT_5f88_a3eb_59e0 addBadFile: penz_2016-03.jsonl
  reason=SimpleException: ERROR while reading /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl: columnName=timestamp not found in longitude,latitude,airTemp,time.
 at gov.noaa.pfel.erddap.dataset.EDDTableFromHttpGet.readFile(EDDTableFromHttpGet.java:370)
  Table.saveAsFlatNc /local/tomcat/content/erddapBigDir/dataset/e0/inbound_NRT_5f88_a3eb_59e0/badFiles.nc1141171690 done. nColumns=3 nRows=2 TIME=3ms
Table of badFiles successfully written. nRows=2
gov.noaa.pfel.erddap.dataset.WaitThenTryAgainException: There was a (temporary?) problem. Wait a minute, then try again. (In a browser, click the Reload button.)
(Cause: com.cohort.util.SimpleException: ERROR while reading /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl: columnName=timestamp not found in longitude,latitude,airTemp,time.)
```

```shell
cat /local/tomcat_shared/content/erddap/data/inbound-NRT/penz/penz_2016-03.jsonl
["stationID","time","latitude","longitude","airTemp","waterTemp","timestamp","author","command"]
["penz","2016-03-30T12:37:55Z",10.1,-150.1,17.23,12.3,1.601565967722E9,"tomG",0]
["penz","2016-03-30T12:37:55Z",10.1,-150.1,17.3,12.3,1.601566535098E9,"tomG",0]
["penz","2016-03-30T13:37:55Z",10.1,-150.1,17.2,12.81,1.601566535098E9,"tomG",0]
["penz","2016-03-30T14:37:55Z",10.1,-150.1,17.72,13.11,1.601566535098E9,"tomG",0]
["penz","2016-03-30T15:37:55Z",10.1,-150.1,17.5,12.793,1.601566535098E9,"tomG",0]
["penz","2016-03-30T16:37:55Z",10.1,-150.1,17.1,1.873,1.601566535098E9,"tomG",0]
["penz","2016-03-30T17:37:55Z",10.1,-150.1,17.3,12.63,1.601566535098E9,"tomG",0]
["penz","2016-03-30T12:37:55Z",10.1,-150.1,17.3,12.3,1.601568148226E9,"tomG",0]
["penz","2016-03-30T13:37:55Z",10.1,-150.1,17.2,12.81,1.601568148226E9,"tomG",0]
["penz","2016-03-30T14:37:55Z",10.1,-150.1,17.72,13.11,1.601568148226E9,"tomG",0]
["penz","2016-03-30T15:37:55Z",10.1,-150.1,17.5,12.793,1.601568148226E9,"tomG",0]
["penz","2016-03-30T16:37:55Z",10.1,-150.1,17.1,1.873,1.601568148226E9,"tomG",0]
["penz","2016-03-30T17:37:55Z",10.1,-150.1,17.3,12.63,1.601568148226E9,"tomG",0]
```

XML snippet used

```XML
<dataset type="EDDTableFromHttpGet" datasetID="inbound_NRT_5f88_a3eb_59e0" active="true">
    <reloadEveryNMinutes>1440</reloadEveryNMinutes>
    <updateEveryNMillis>-1</updateEveryNMillis>
    <fileDir>/local/tomcat_shared/content/erddap/data/inbound-NRT/</fileDir>
    <fileNameRegex>.*\.jsonl</fileNameRegex>
    <recursive>true</recursive>
    <pathRegex>.*</pathRegex>
    <metadataFrom>last</metadataFrom>
    <sortedColumnSourceName></sortedColumnSourceName>
    <sortFilesBySourceNames>stationID, time</sortFilesBySourceNames>
    <fileTableInMemory>true</fileTableInMemory>
    <accessibleViaFiles>true</accessibleViaFiles>
    <addAttributes>
        <att name="cdm_data_type">Point</att>
        <att name="Conventions">COARDS, CF-1.6, ACDD-1.3</att>
        <att name="creator_name">BODC</att>
        <att name="creator_type">institution</att>
        <!--att name="cdm_timeseries_variables">stationID, longitude, latitude</att-->
        <att name="subsetVariables">station_id, longitude, latitude</att>
        <att name="httpGetDirectoryStructure">stationID/2months</att>
        <att name="httpGetKeys"> -- keys here --</att>
        <att name="httpGetRequiredVariables">stationID, time</att>
        <att name="infoUrl">???</att>
        <att name="institution">BODC</att>
        <att name="keywords">air, airTemp, author, bodc, british, centre, command, creamt, data, identifier, latitude, longitude, oceanographic, station, stationID, temperature, tester, time, timestamp, water, waterTemp</att>
        <att name="license">[standard]</att>
        <att name="sourceUrl">(local files)</att>
        <att name="standard_name_vocabulary">CF Standard Name Table v55</att>
        <att name="subsetVariables">stationID</att>
        <att name="summary">a test of the from http method for us by NOC and British Oceanographic Data Centre (BODC)</att>
        <att name="testOutOfDate">now-1day</att>
        <att name="title">creamT tester</att>
    </addAttributes>


    <dataVariable>
        <sourceName>stationID</sourceName>
 <destinationName>stationID</destinationName>
        <dataType>String</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="long_name">Station ID</att>
            <!--att name="cf_role">timeseries_id</att-->
        </addAttributes>
    </dataVariable>


    <dataVariable>
        <sourceName>time</sourceName>
        <destinationName>time</destinationName>
        <dataType>String</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="long_name">Time</att>
            <att name="standard_name">time</att>
            <att name="time_precision">1970-01-01T00:00:00Z</att>
            <att name="units">yyyy-MM-dd&#39;T&#39;HH:mm:ss&#39;Z&#39;</att>
        </addAttributes>
    </dataVariable>


    <dataVariable>
        <sourceName>latitude</sourceName>
        <destinationName>latitude</destinationName>
        <dataType>double</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="colorBarMaximum" type="double">90.0</att>
            <att name="colorBarMinimum" type="double">-90.0</att>
            <att name="long_name">Latitude</att>
            <att name="missing_value" type="double">NaN</att>
            <att name="standard_name">latitude</att>
            <att name="units">degrees_north</att>
        </addAttributes>
   </dataVariable>


    <dataVariable>
        <sourceName>longitude</sourceName>
        <destinationName>longitude</destinationName>
        <dataType>double</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="colorBarMaximum" type="double">180.0</att>
            <att name="colorBarMinimum" type="double">-180.0</att>
            <att name="long_name">Longitude</att>
            <att name="missing_value" type="double">NaN</att>
            <att name="standard_name">longitude</att>
            <att name="units">degrees_east</att>
        </addAttributes>
    </dataVariable>
 
    <dataVariable>
        <sourceName>airTemp</sourceName>
        <destinationName>airTemp</destinationName>
        <dataType>float</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="long_name">Air Temp</att>
            <att name="missing_value" type="float">NaN</att>
            <att name="units">degrees C</att>
        </addAttributes>
    </dataVariable>
  

  <dataVariable>
        <sourceName>waterTemp</sourceName>
        <destinationName>waterTemp</destinationName>
        <dataType>float</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="long_name">Water Temp</att>
            <att name="missing_value" type="float">NaN</att>
            <att name="units">degrees C</att>
        </addAttributes>
    </dataVariable>

    <dataVariable>
       <sourceName>timestamp</sourceName>
        <destinationName>timestamp</destinationName>
        <dataType>double</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="long_name">Timestamp</att>
            <att name="missing_value" type="double">NaN</att>
            <att name="time_precision">1970-01-01T00:00:00.000Z</att>
            <att name="units">seconds since 1970-01-01T00:00:00Z</att>
        </addAttributes>
    </dataVariable>


    <dataVariable>
        <sourceName>author</sourceName>
        <destinationName>author</destinationName>
        <dataType>String</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="ioos_category">Identifier</att>
            <att name="long_name">Author</att>
        </addAttributes>
    </dataVariable>


    <dataVariable>
        <sourceName>command</sourceName>
        <destinationName>command</destinationName>
        <dataType>byte</dataType>
        <!-- sourceAttributes>
                     </sourceAttributes -->
        <addAttributes>
            <att name="flag_meanings">insert delete</att>
            <att name="flag_values" type="byteList">0 1</att>
            <att name="ioos_category">Other</att>
            <att name="long_name">Command</att>
            <att name="missing_value" type="byte">127</att>
        </addAttributes>
    </dataVariable>


</dataset>
```

**Comment by BobSimons** [comment]
Thanks for providing all the background information. 
You are the first or one of the first users of the EDDTableFromHttpGet system, so it isn't too surprising there are some problems with it.
Yes, I see something is wrong  I will investigate, work to fix it, and get back to you.

In the meantime, as a workaround,
if all of the 'command' values are 0=add, 
you can create a 2nd dataset with a datasets.xml chunk which is almost identical to this dataset's, except:
* change the dataset type to EDDTableFromJsonlCSVFiles
* change the datasetID
* remove the 3 httpGet... global attributes
Then users can read data from those datafiles (and you can add data to the files if you stick to the way EDDTableFromHttpGet adds data).
But users won't be able to use the special features of EDDTableFromHttpGet.

I hope that helps.

**Comment by emmerbodc** [comment]
Thanks for investigating this. 

We'll look into this.

**Comment by BobSimons** [comment]
I've made the fixes and a few related small changes. 
The changes will be in the next release, hopefully very soon -- I'm just finishing up a few tests and documentation changes for this release (but I have a conference next week).
Thanks again for trying EDDTableFromHttpGet and reporting the problems.

**Comment by emmerbodc** [comment]
Great thanks. We will let you know how we get on.

**Comment by BobSimons** [comment]
This has been fixed. The fix is available in the newly released ERDDAP v2.10.
Thanks again for reporting this.

**Comment by BobSimons** [comment]
This was fixed in ERDDAP v2.10.
Thanks again for reporting this.
